packages/react-reconciler/src/ReactMemoCache.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {MemoCache} from 'shared/ReactTypes';
+import type {Dispatcher} from './ReactInternalTypes';
+
+import ReactSharedInternals from 'shared/ReactSharedInternals';
+const {ReactCurrentDispatcher} = ReactSharedInternals;
+
+export const MEMO_CACHE_UNSET_SENTINEL: symbol = Symbol.for(
+  'react.usememocache_sentinel',
+);
+
+/**
+ * Implementation of `useMemoCache()` return type
+ */
+export default class MemoCacheImpl implements MemoCache {
```

Comment:
React runtime doesn't use classes. Maybe it's okay to start using them now, but I think there's concerns about random overhead generated by transforms to be fully compatible.

Maybe this could use `Object.create` instead?